prow/cmd/branchprotector/protect_test.go

```diff
@@ -365,317 +393,244 @@ func split(branch string) (string, string, string) {
-		for _, b := range tc.branches {
-			org, repo, branch := split(b)
-			k := org + "/" + repo
-			branches[k] = append(branches[k], github.Branch{Name: branch})
-			r := repos[org]
-			if r == nil {
-				repos[org] = make(map[string]bool)
+		t.Run(tc.name, func(t *testing.T) {
+			repos := map[string]map[string]bool{}
+			branches := map[string][]github.Branch{}
+			for _, b := range tc.branches {
+				org, repo, branch := split(b)
+				k := org + "/" + repo
+				branches[k] = append(branches[k], github.Branch{Name: branch})
+				r := repos[org]
+				if r == nil {
+					repos[org] = make(map[string]bool)
+				}
+				repos[org][repo] = true
 			}
-			repos[org][repo] = true
-		}
-		fc := FakeClient{
-			deleted:  make(map[string]bool),
-			contexts: make(map[string][]string),
-			pushers:  make(map[string][]string),
-			branches: branches,
-			repos:    make(map[string][]github.Repo),
-		}
-		for org, r := range repos {
-			for rname := range r {
-				fc.repos[org] = append(fc.repos[org], github.Repo{Name: rname, FullName: org + "/" + rname})
+			fc := FakeClient{
+				deleted:  make(map[string]bool),
+				branches: branches,
+				repos:    make(map[string][]github.Repo),
+			}
+			for org, r := range repos {
+				for rname := range r {
+					fc.repos[org] = append(fc.repos[org], github.Repo{Name: rname, FullName: org + "/" + rname})
+				}
 			}
-		}
-
-		p := Protector{
-			client:         &fc,
-			cfg:            &tc.config,
-			errors:         Errors{},
-			updates:        make(chan Requirements),
-			done:           make(chan []error),
-			completedRepos: make(map[string]bool),
-		}
-		go func() {
-			p.Protect()
-			close(p.updates)
-		}()
 
-		var actual []Requirements
-		for r := range p.updates {
-			actual = append(actual, r)
-		}
-		errors := p.errors.errs
-		if len(errors) != tc.errors {
-			t.Errorf("%s: actual errors %d != expected %d", tc.name, len(errors), tc.errors)
-		}
-		switch {
-		case len(actual) != len(tc.expected):
-			t.Errorf("%s: actual updates %v != expected %v", tc.name, actual, tc.expected)
-		default:
-			for _, a := range actual {
-				found := false
-				for _, e := range tc.expected {
-					if e.Org == a.Org && e.Repo == a.Repo && e.Branch == a.Branch {
-						found = true
-						switch {
-						case e.Protect != a.Protect:
-							t.Errorf("%s: %s/%s=%s actual protect %t != expected %t", tc.name, e.Org, e.Repo, e.Branch, a.Protect, e.Protect)
-						case len(e.Contexts) != len(a.Contexts), len(e.Contexts) > 0 && !reflect.DeepEqual(e.Contexts, a.Contexts):
-							t.Errorf("%s: %s/%s=%s actual contexts %v != expected %v", tc.name, e.Org, e.Repo, e.Branch, a.Contexts, e.Contexts)
-						case len(e.Pushers) != len(a.Pushers), len(e.Pushers) > 0 && !reflect.DeepEqual(e.Pushers, a.Pushers):
-							t.Errorf("%s: %s/%s=%s actual pushers %v != expected %v", tc.name, e.Org, e.Repo, e.Branch, a.Pushers, e.Pushers)
-						}
-						break
+			var cfg config.Config
+			err := yaml.Unmarshal([]byte(tc.config), &cfg)
```

Comment:
Not a fan of this at all, before you have methods that operate on Go structs so you had provided them with the data they needed and checked the output. Now, you added YAML marshaling and unmarshalling to the "unit" under test. If I want to rename or refactor something in my IDE, none of these opaque strings will be changed.

Reply:
This is a better way to write this test. What matters is that a given yaml input produces the same github client calls.